dbms/src/DataStreams/NullableAdapterBlockInputStream.cpp

```diff
@@ -92,37 +48,95 @@ void NullableAdapterBlockInputStream::buildActions(
     const Block & in_sample,
     const Block & out_sample)
 {
-    size_t in_size = in_sample.columns();
+    size_t out_size = out_sample.columns();
 
-    if (out_sample.columns() != in_size)
+    /// TODO: Maybe remove this when have default value in the column.
+    if (in_sample.columns() != out_size)
         throw Exception("Number of columns in INSERT SELECT doesn't match", ErrorCodes::NUMBER_OF_COLUMNS_DOESNT_MATCH);
 
-    actions.reserve(in_size);
-    rename.reserve(in_size);
+    actions.reserve(out_size);
+
+    bool buildWithName = true;
 
-    for (size_t i = 0; i < in_size; ++i)
+    for (size_t i = 0; i < out_size; ++i)
+    {
+        const auto &out_elem = out_sample.getByPosition(i);
+        if (!in_sample.has(out_elem.name))
```

Comment:
Unlike maser use of position to match, only position is used when name cannot match.

Reply:
I changed ` INSERT INTO test.insert_fewer_columns_2 SELECT * FROM test.insert_fewer_columns;`  results.
before: `Row 1 : 0	 1`  `Row 2 : 0  2` 
after: `Row 1: 1  0` `Row 2 : 2  0`

Reply:
And `https://github.com/yandex/ClickHouse/issues/390#issuecomment-299881474` not pass in the master

Reply:
Ok.

Reply:
Switching between matching by positions and by names is too implicit.
I think it may lead to unexpected behaviour - when the user change query a little it will lead to completely different result.

It's better to make it explicitly switchable by setting.

Reply:
👍 This is great, But, `INSERT INTO test(a,b,c) SELECT b,c,a FROM test_b` always confusing to report wrong when matching by positions

Reply:
This query looks like the user really wants to map by positions :)
Instead they will write `INSERT INTO test (a, b, c) SELECT a, b, c FROM test_b`
or `INSERT INTO test (a, b, c) SELECT expr1 as b, expr2 as c, expr3 as a FROM test_b`.

Reply:
OK,I agree : )